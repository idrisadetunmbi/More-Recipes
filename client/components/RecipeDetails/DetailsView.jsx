import React from 'react';
import PropTypes from 'prop-types';
import UserIcon from './avatar_img.png';
import { LoaderWithComponent } from '../reusables';

const DetailsView = (props) => {
  const { userVoteStatuses } = props;
  const userUpvoted = userVoteStatuses && userVoteStatuses.upvoted;
  const userDownvoted = userVoteStatuses && userVoteStatuses.downvoted;
  const userFavorited = userVoteStatuses && userVoteStatuses.favorited;

  const userIsSignedIn = () => {
    const { user, history } = props;
    // if user is not signed in, redirect user to sign in
    if (!user.token) {
      history.push('/signin', {
        modal: true,
        previousLocation: history.location.pathname,
      });
      return false;
    }
    return true;
  };

  const { recipe, user } = props;
  return (
    <div className="recipe-details-component">
      <div className="container-section container">
        <div className="row" id="recipe-info">
          {/* recipe images and action buttons start */}
          <div className="col s12 l4 images-section">
            <h5 className="title">{recipe.title}</h5>
            <div className="divider" />
            {/* recipe author section */}
            <div id="recipe-author-section">
              {
                recipe.author.imageUrl ?
                  <img
                    src={recipe.author.imageUrl}
                    alt=""
                    className="responsive-img circle"
                  /> :
                  <i className="material-icons">account_circle</i>
              }
              <div>
                <span>Authored by</span>
                <p>
                  {props.userOwnsRecipe() ? 'You' : recipe.author.username}
                </p>
              </div>
            </div>
            <div>
              <div>
                <div className="recipe-main-img">
                  <img src={`${recipe.images[0]}`} alt="" />
                </div>
                <div id="thumbnails-row" className="row">
                  {
                    // TODO: replace image sources
                    (recipe.images.slice(1).map((image, i) =>
                      (<img
                        style={{ marginRight: `${!(i % 2) && '2%'}` }}
                        className="col s6 thumbnails"
                        src={image}
                        alt=""
                      />))
                    )
                  }
                </div>
              </div>

              {/* TODO: wire actions to onClicks of the action buttons */}
              <div>
                <div className="vote-actions">
                  <a
                    style={{ marginLeft: '0' }}
                    className={`btn waves-ripple ${userUpvoted && 'checked'}`}
                    disabled={props.userOwnsRecipe()}
                    onClick={() => userIsSignedIn() &&
                      props.recipeVoteAction('upvote', recipe.id)}
                  >
                    <i className="material-icons">thumb_up</i>
                    <span>{recipe.upvotes}</span>
                  </a>
                  <a
                    className={`btn waves-ripple ${userDownvoted && 'checked'}`}
                    disabled={props.userOwnsRecipe()}
                    onClick={() => userIsSignedIn() &&
                      props.recipeVoteAction('downvote', recipe.id)}
                  >
                    <i className="material-icons">thumb_down</i>
                    <span>{recipe.downvotes}</span>
                  </a>
                  <a
                    className={`btn waves-ripple ${userFavorited && 'checked'}`}
                    disabled={props.userOwnsRecipe()}
                    onClick={() => userIsSignedIn() &&
                      props.recipeVoteAction('favorite', recipe.id)}
                  >
                    <i className="material-icons">favorite</i>
                    <span>{recipe.favorites}</span>
                  </a>
                </div>
              </div>
            </div>
          </div>
          {/* recipe images and action buttons end */}

          <div className="col s12 offset-l1 l6 description-section">
            <div>
              <h5>Description</h5>
              <p>{recipe.description}</p>
            </div>
            <div className="divider" />
            <div className="ingredients">
              <h5>Ingredients</h5>
              <ul>
                {
                  recipe.ingredients.split('\n')
                    .map((ingredient, i) =>
                      (
                        <li>
                          <p>
                            <i className="material-icons">chevron_right</i>
                            {ingredient}
                          </p>
                        </li>
                      ))
                }
              </ul>
            </div>
            <div className="divider" />
            <div className="directions">
              <h5>Directions</h5>
              <ol>
                {
                  recipe.directions.split('\n').map(direction =>
                    <li><p>{direction}</p></li>)
                }
              </ol>
            </div>
          </div>
        </div>
        {/* recipe details end */}

        {/* review section */}
        <div className="divider" />
        <div id="review-section">
          <h5>Reviews</h5>
          <div className="row">
            <div id="review-form" className="col offset-m1 m10 s12 offset-l3 l6">
              <textarea
                value={props.reviewText}
                onChange={props.reviewOnChange}
                className="materialize-textarea"
                placeholder="Add a review"
              />
              <button onClick={props.reviewSubmit} className="btn btn-small">Submit</button>
            </div>

            {
              <LoaderWithComponent
                showLoader={!props.reviews}
                component={
                  props.reviews && props.reviews.length ?
                    props.reviews.map(review => (
                      <Review review={review} key={review.id} />)) :
                    <p id="no-review-text" className="col l6 offset-l3">
                      This Recipe currently has no reviews
                    </p>
                }
              />
            }
          </div>
        </div>
      </div>

      {
        user.token && props.userOwnsRecipe() &&
        (
          <div className="fixed-action-btn">
            <a className="btn-floating btn-large">
              <i className="large material-icons">menu</i>
            </a>
            <ul>
              <li>
                <a
                  onClick={() => props.recipeAction('delete', recipe.id)}
                  className="btn-floating red"
                >
                  <i className="material-icons">delete</i>
                </a>
              </li>
              <li>
                <a
                  onClick={props.toggleViewMode}
                  className="btn-floating yellow darken-1"
                >
                  <i className="material-icons">edit</i>
                </a>
              </li>
            </ul>
          </div>
        )
      }
    </div>
  );
};

const Review = (props) => {
  const reviewer = props.review.user;
  const date = new Date(props.review.createdAt);
  const month = date.toLocaleString('en-us', { month: 'short' });
  const day = date.getDate();

  return (
    <div className="review col offset-m1 m10 s12 offset-l3 l6">
      <img
        className="circle"
        src={reviewer.imageUrl || UserIcon}
        width="35"
        height="35"
        alt="reviewers icon"
      />
      <div className="review-details">
        <span className="reviewers-name">{reviewer.username}</span>
        <span className="date">{`${month} ${day}`}</span>
      </div>
      <div className="divider" />
      <p className="review-text">
        {props.review.content}
      </p>
    </div>
  );
};

DetailsView.propTypes = {
  recipe: PropTypes.shape().isRequired,
  user: PropTypes.shape().isRequired,
  recipeAction: PropTypes.func.isRequired,
  recipeVoteAction: PropTypes.func.isRequired,
  reviews: PropTypes.arrayOf(PropTypes.object),
  reviewText: PropTypes.string.isRequired,
  reviewOnChange: PropTypes.func.isRequired,
  reviewSubmit: PropTypes.func.isRequired,
  toggleViewMode: PropTypes.func.isRequired,
  userVoteStatuses: PropTypes.shape(),
  userOwnsRecipe: PropTypes.func.isRequired,
};

DetailsView.defaultProps = {
  reviews: undefined,
  userVoteStatuses: undefined,
};

Review.propTypes = {
  review: PropTypes.shape().isRequired,
};

export default DetailsView;
