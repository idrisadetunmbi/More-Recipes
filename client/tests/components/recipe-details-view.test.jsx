import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

import DetailsView from '../../components/RecipeDetails/DetailsView';

const recipeDetailsSetUp = (dynamicProps) => {
  const props = {
    recipe: {
      id: 1000,
      title: 'A great recipe',
      description: 'a great recipe',
      ingredients: 'ingredients1\ningredients2',
      directions: 'direction1\ndirection2',
      images: ['', ''],
      author: {
        imageUrl: 'imageUrl',
        username: 'username',
      },
    },
    user: {
      token: 'token',
    },
    recipeAction: jest.fn(),
    recipeVoteAction: jest.fn(),
    reviews: [{
      user: { username: 'testuser', imageUrl: 'testimageUrl' },
      content: 'reviewcontent',
      createdAt: Date.now(),
    }],
    userOwnsRecipe: () => true,
    reviewText: '',
    reviewOnChange: jest.fn(),
    reviewSubmit: jest.fn(),
    toggleViewMode: jest.fn(),
    userVoteStatuses: {
      upvoted: true,
      downvoted: true,
      favorited: true,
    },
    history: {
      push: jest.fn(),
      location: {},
    },
    ...dynamicProps,
  };
  const component = mount(<DetailsView {...props} />);
  return {
    props,
    component,
    elements: {
      recipeTitle: component.find('h5.title'),
      upvoteBtn: component.find('a#upvote-btn'),
      downvoteBtn: component.find('a#downvote-btn'),
      favoriteBtn: component.find('a#favorite-btn'),
      deleteBtn: component.find('a.btn-floating.red'),
    },
  };
};

describe('RecipeDetails - DetailsView component', () => {
  test('matches snapshot', () => {
    const props = {
      userOwnsRecipe: jest.fn(),
      user: { data: {} },
      recipe: {
        author: {},
        images: { slice: () => [] },
        ingredients: { split: () => [], replace: jest.fn() },
        directions: { split: () => [], replace: jest.fn() },
      },
    };
    const component = renderer.create(<DetailsView {...props} />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  test('renders title correctly', () => {
    const { elements: { recipeTitle }, props } = recipeDetailsSetUp();
    expect(recipeTitle).toHaveText(props.recipe.title);
  });

  test('upvote button calls voteAction function', () => {
    const { elements: { upvoteBtn }, props: { recipeVoteAction } } =
      recipeDetailsSetUp();
    upvoteBtn.simulate('click');
    expect(recipeVoteAction).toHaveBeenCalled();
  });

  test('downvote button calls voteAction function', () => {
    const { elements: { downvoteBtn }, props: { recipeVoteAction, recipe } } =
      recipeDetailsSetUp();
    downvoteBtn.simulate('click');
    expect(recipeVoteAction).toHaveBeenCalledWith('downvote', recipe.id);
  });

  test('favorite button calls voteAction function', () => {
    const { elements: { favoriteBtn }, props: { recipeVoteAction, recipe } } =
      recipeDetailsSetUp();
    favoriteBtn.simulate('click');
    expect(recipeVoteAction).toHaveBeenCalledWith('favorite', recipe.id);
  });

  test('delete button calls recipeAction function', () => {
    const { elements: { deleteBtn }, props: { recipeAction, recipe } } =
      recipeDetailsSetUp();
    deleteBtn.simulate('click');
    expect(recipeAction).toHaveBeenCalledWith('delete', recipe.id);
  });

  test('calls history.push function when user is not signed in', () => {
    const { elements: { upvoteBtn }, props: { history: { push } } } =
      recipeDetailsSetUp({ user: { token: '' } });
    upvoteBtn.simulate('click');
    expect(push).toHaveBeenCalled();
  });
});
