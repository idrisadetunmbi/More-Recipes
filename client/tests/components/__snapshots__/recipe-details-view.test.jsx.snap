// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecipeDetails - DetailsView component matches snapshot 1`] = `
<div
  className="recipe-details-component"
>
  <div
    className="container-section container"
  >
    <div
      className="row"
      id="recipe-info"
    >
      <div
        className="col s12 l4 images-section"
      >
        <h5
          className="title"
        />
        <div
          className="divider"
        />
        <div
          id="recipe-author-section"
        >
          <i
            className="material-icons"
          >
            account_circle
          </i>
          <div>
            <span>
              Authored by
            </span>
            <p />
          </div>
        </div>
        <div>
          <div>
            <div
              className="recipe-main-img"
            >
              <img
                alt=""
                src="undefined"
              />
            </div>
            <div
              className="row"
              id="thumbnails-row"
            />
          </div>
          <div>
            <div
              className="vote-actions"
            >
              <a
                className="btn waves-ripple undefined"
                disabled={undefined}
                id="upvote-btn"
                onClick={[Function]}
                style={
                  Object {
                    "marginLeft": "0",
                  }
                }
              >
                <i
                  className="material-icons"
                >
                  thumb_up
                </i>
                <span />
              </a>
              <a
                className="btn waves-ripple undefined"
                disabled={undefined}
                id="downvote-btn"
                onClick={[Function]}
              >
                <i
                  className="material-icons"
                >
                  thumb_down
                </i>
                <span />
              </a>
              <a
                className="btn waves-ripple undefined"
                disabled={undefined}
                id="favorite-btn"
                onClick={[Function]}
              >
                <i
                  className="material-icons"
                >
                  favorite
                </i>
                <span />
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        className="col s12 offset-l1 l6 description-section"
      >
        <div>
          <h5>
            Description
          </h5>
          <p />
        </div>
        <div
          className="divider"
        />
        <div
          className="ingredients"
        >
          <h5>
            Ingredients
          </h5>
          <ul />
        </div>
        <div
          className="divider"
        />
        <div
          className="directions"
        >
          <h5>
            Directions
          </h5>
          <ol />
        </div>
      </div>
    </div>
    <div
      className="divider"
    />
    <div
      id="review-section"
    >
      <h5>
        Reviews
      </h5>
      <div
        className="row"
      >
        <div
          className="col offset-m1 m10 s12 offset-l3 l6"
          id="review-form"
        >
          <textarea
            className="materialize-textarea"
            onChange={undefined}
            placeholder="Add a review"
            value={undefined}
          />
          <button
            className="btn btn-small"
            onClick={undefined}
          >
            Submit
          </button>
        </div>
        <div
          className="preloader"
        >
          <div
            className="preloader-wrapper big active"
          >
            <div
              className="spinner-layer"
            >
              <div
                className="circle-clipper left"
              >
                <div
                  className="circle"
                />
              </div>
              <div
                className="gap-patch"
              >
                <div
                  className="circle"
                />
              </div>
              <div
                className="circle-clipper right"
              >
                <div
                  className="circle"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
